### PR TITLE
Strictly align our errors with uber-go guidelines

### PR DESCRIPTION
**Pull request by omerfirmak**
https://github.com/uber-go/guide/blob/master/style.md#errors

### CODE REVIEW

**Comment by omerfirmak** [comment]
I am not opinionated at all on this matter. So please make any changes (including dropping commits) you deem necessary before approving, instead of leaving requests.

**Comment by omerfirmak** [comment]
Rebased

File: core/transaction.go
Comment (jelilat) [inline core/transaction.go:1]: Random. I think this should be `not matched,  dynamic error message`

File: core/trie/node_test.go
Comment (jelilat) [inline core/trie/node_test.go:1]: We probably shouldn't be adding unrelated changes

**Review by jelilat — COMMENTED** [review]
Random. I think this should be `not matched, dynamic error message`

File: core/transaction.go [hunk 1]
```diff
@@ -233,6 +223,10 @@ var (
 	deployAccountFelt = new(felt.Felt).SetBytes([]byte("deploy_account"))
 )
 
+func errInvalidTransactionVersion(t Transaction, version *felt.Felt) error {
+	return fmt.Errorf("invalid Transaction (type: %v) verion: %v", reflect.TypeOf(t), version.Text(felt.Base10))
+}
+
```
Comment (IronGauntlets) [inline core/transaction.go:227]: let's create the following error (please see error naming section in uber-go):

```go
type invalidTransactionVersionError struct {	
	t Transaction	
	v string	
}	

func (e ErrInvalidTransactionVersion) Error() string {	
	return fmt.Sprintf(""invalid Transaction (type: %v) verion: %v", reflect.TypeOf(t), version.Text(felt.Base10))
}
```

And add a test for an unsupported version which uses require/assert's `ErrorIs`/`ErrorAs`.

File: core/transaction.go [hunk 2]
```diff
@@ -320,23 +312,23 @@ func deployAccountTransactionHash(d *DeployAccountTransaction, n utils.Network)
 			d.Nonce,
 		), nil
 	}
-	return nil, ErrInvalidTransactionVersion{d, d.Version.Text(felt.Base10)}
+	return nil, errInvalidTransactionVersion(d, d.Version)
 }
 
-type ErrCantVerifyTransactionHash struct {
+type CantVerifyTransactionHashError struct {
```
Comment (jelilat) [inline core/transaction.go:318]: @IronGauntlets thought we wanted to follow the convention `Err...`? Uber Guidelines also suggested the same https://github.com/uber-go/guide/blob/master/style.md#error-naming

Comment (IronGauntlets) [reply core/transaction.go:318]: yeah you are right I forgot to make it private

**Comment by IronGauntlets** [comment]
Please replace `ErrorEqual` with `ErrorAs`/`ErrorIs` where appropriate, this will fulfill the matching condition allowing us to create private and public package-level errors.

**Comment by omerfirmak** [comment]
> Please replace `ErrorEqual` with `ErrorAs`/`ErrorIs` where appropriate, this will fulfill the matching condition allowing us to create private and public package-level errors.

I dont think matching them in tests fulfills the requirement.

File: core/transaction.go (hunk 1 above)
Comment (omerfirmak) [reply core/transaction.go:227]: We shouldn't create typed errors unless we are explicitly handling them somewhere. Checking them in tests does not warrant a typed error since they are not really "handling" the error.

**Review by IronGauntlets — APPROVED** [review]
